lib/trino-parquet/src/main/java/io/trino/parquet/reader/flat/FlatColumnReader.java

```diff
@@ -372,4 +345,202 @@ private static void throwEndOfBatchException(int remainingInBatch)
     {
         throw new ParquetDecodingException(format("Corrupted Parquet file: extra %d values to be consumed when scanning current batch", remainingInBatch));
     }
+
+    private ValuesBuffer<BufferType> createValuesBuffer(int batchSize)
+    {
+        // Parquet writer may choose to fall back to a non-dictionary encoding after starting with dictionary encoding if
+        //   1. If the size of the dictionary exceeds a threshold (1MB for parquet-mr by default).
+        //   2. Number of dictionary entries exceeds a threshold (Integer.MAX_VALUE for parquet-mr by default).
+        // Trino dictionary blocks are produced only when the entire column chunk is dictionary encoded
+        if (pageReader.hasOnlyDictionaryEncodedPages()) {
+            requireNonNull(dictionaryDecoder, "dictionaryDecoder is null");
+            // Filtering of parquet pages using column indexes may result in the total number of values read from the
+            // column chunk being lower than the size of the dictionary
+            if (pageReader.getTotalValueCount() > dictionaryDecoder.getDictionarySize()) {
+                return new DictionaryValuesBuffer<>(dictionaryDecoder, batchSize);
+            }
+        }
+        return new DataValuesBuffer<>(columnAdapter, batchSize);
+    }
+
+    private interface ValuesBuffer<T>
```

Comment:
How is this different from ColumnAdapter? There seems to be some overlap.

Reply:
This owns state (buffer containing values for current batch) while `ColumnAdapter` doesn't.
It also keeps track of count of nulls in the current batch and uses that to produce RLE block of nulls if all values are nulls or a non-block for a nullable column if the current batch had no nulls.
It's not doing anything type specific whereas `ColumnAdapter` is mainly about things that need to be done differently for various types.